scipy/stats/sampling.py

```diff
@@ -49,3 +70,1262 @@
+        self._pdf = lambda x: pdf(x, *args)
+
+    def pdf(self, x):
+        return self._pdf(x)
+
+
+class FastGeneratorInversion:
+    """
+    Fast sampling by numerical inversion of the CDF for a large class of
+    continuous distributions in `scipy.stats`.
+
+    Parameters
+    ----------
+    distname : str
+        Name of the distribution in `scipy.stats`.
+    args : tuple of floats or empty tuple, optional
+        The shape parameters (without `loc` and `scale`) of the distribution.
+        Important note: Only a single set of shape parameters is allowed, e.g.,
+        for the Gamma distribution with shape parameter `p`, `p` has to be a
+        float, and for the beta distribution with shape parameters (a, b), both
+        a and b have to be floats. The default is (), i.e., the distribution
+        has no shape parameters.
+    domain : tuple of floats, optional
+        If one wishes to sample from a truncated/conditional distribution,
+        the domain has to be specified.
+        The default is None. In that case, the random variates are not
+        truncated, and the domain is inferred from the support of the
+        distribution.
+    loc : float, optional
+        The location parameters. Note that only floats and no arrays are
+        allowed. The default is 0.
+    scale : float, optional
+        The scale parameter. Note that an array of floatsis not allowed.
+        The default is 1.
+    ignore_shape_range : boolean, optional.
+        If False, shape parameters that are outside of the valid range
+        of values to ensure that the numerical accuracy (see Notes) is
+        high, raise a ValueError. If True, any shape parameters that are valid
+        for the distribution are accepted. This can be useful for testing.
+        The default is False.
+        If set to True, warnings are not suppresed that might occur when an
+        instance of the class is created.
+    random_state : {None, int, `numpy.random.Generator`,
+                        `numpy.random.RandomState`}, optional
+
+            A NumPy random number generator or seed for the underlying NumPy
+            random number generator used to generate the stream of uniform
+            random numbers.
+            If `random_state` is None, it uses ``self.random_state``.
+            If `random_state` is an int,
+            ``np.random.default_rng(random_state)`` is used.
+            If `random_state` is already a ``Generator`` or ``RandomState``
+            instance then that instance is used.
+
+    Attributes
+    ----------
+    random_state : {`numpy.random.Generator`, `numpy.random.RandomState`}
+        The random state used in relevant methods like `rvs` (unless
+        another `random_state` is passed as an argument to these methods).
+
+    Methods
+    -------
+    cdf
+    evaluate_error
+    isf
+    logpdf
+    pdf
+    ppf
+    ppf_fast
+    qrvs
+    rvs
+    set_loc_scale
+    support
+    sf
```

Comment:
I am concerned about the API. What is the scope of this class? As it is, as a users, this would confuse me with distributions, frozen dists and this.

I really think we should focus on the sampling part, as the description suggests. i.e. here only propose `rvs, qrvs, evaluate_error`. If needed internally, others things should be private. 

For `rvs, qrvs` ideally I would prefer to only have a single method and set this during the init. But since we already have this API elsewhere I suppose it's ok. Another though is that I am not sure if its useful to keep a non QMC option.

Reply:
I think pdf, cdf, ppf are important to have, otherwise you always need to create another object from rv_continuous to have access to these methods. For example, if you want to create a histogram, you need to do something like

```
x = np.linspace(0, 1, 100)
rng1 = FastGeneratorInversion('beta', (1.5, 2.5))
rng2 = stats.beta(1.5, 2.5)
r = rng1.rvs(size=1000)
y = rng2.pdf(x)
...
```

and if you use truncation, we do not support that in rv_continuous yet, so the user would manually have to compute the pdf/cdf

Reply:
I would agree to removing the following methods: `isf`, `logpdf`, `sf`

`sf` and `isf` might be required in some situations but probably not very often, `logpdf` might be most relevant for MLE and `fit` is not supported anyways. this would already simplify the API quite a bit while keeping the key functions that are required in the context of sampling

Reply:
I also discussed this topic with Matt. There are mainly four options:

1. implement all methods that we have in a frozen object for consistency, e.g., `moment`, `logpdf`, etc including support for truncation
2. focus on sampling including basic support for `pdf` and `cdf` (to create histograms, run a goodness-of-fit test)
3. focus on the sampling, i.e., keep only `rvs`, `qrvs`, `ppf_fast`, `evaluate_error`. `pdf`/`cdf` could be added later if  requested by users.
4. integrate the approach directly into `sampling.NumericalInversePolynomial`, i.e., if a frozen SciPy dist is passed, the settings proposed in this PR are used to create the generator

From my point of view, 1. is too much extra work and there is too much overlap with the strategic API that it intended to support truncation.
Since there is not enough support for 2., I would propose option 3.

After thinking a bit more about option 4, I don't think it is a good idea for various reasons:
- both the documentation and implementation will become confusing with many case distinctions that describe the different behaviors. this PR was intended to make sampling methods more accessible to users who are not experts in random variate generation, while keeping full flexibility in the original `sampling.NumericalInversePolynomial`. Also from this perspective, it is easier to keep them separated.
- I would prefer to minimize changes to the UNU.RAN integration. for example, `sampling.NumericalInversePolynomial` does not even have a separate `rvs` method, all generators use a generic rvs methdod in `stats/_unuran/unuran_wrapper.pyx` which is nice from a consistency point of view.